### PR TITLE
Add async script RFC

### PR DESCRIPTION
**Pull request by hdevalence**
Rendered: https://github.com/ZcashFoundation/zebra/blob/465cd03fb311a28cff1db914af8c69a6e998074c/book/src/dev/rfcs/XXXX-asynchronous-script-verification.md

### CODE REVIEW

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (yaahc) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md]: Needs clarification / more detailed explanation of what "checks them against the hashmap keys" means.

As far as I can tell this is a helper function for calling `respond` for each TransparentOutput that has a corresponding `OutPoint` already in the list of PendingUtxos, is this correct?

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (yaahc) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:149]: Do we need to concern ourselves with how the state struct is managing the commited state? I think from the perspective of the pending utxos we might be able to simplify this to "in the state" and "not in the state" without unnecessarily forcing the lookups to serialize in a way that hurts the common case with old utxos.

**Review by yaahc — COMMENTED** [review]
Looks good, I just have a couple of questions asking for clarification.

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (hdevalence) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md]: Yes, exactly, I can clarify that it's a helper function.

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (hdevalence) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:149]: This is supposed to be relatively agnostic to how the actual state structure is organized, the point is just to divide between data that's in sled and data that isn't; we have to check both.

Comment (yaahc) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:149]: > This is supposed to be relatively agnostic to how the actual state structure is organized, the point is just to divide between data that's in sled and data that isn't; we have to check both.

Agreed. But the fact that we don't care how the state structure is organized implies that it is irrelevant here and shouldn't be mentioned as part of this RFC. Checking both is the same thing as "checking the state structure".

It might be worth adding a definition for "The State Structure" or w/e piece of the system inside the state service handles inserting and getting from the state as a whole.

Comment (hdevalence) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:149]: Hmm, but it does seem important to name the difference, because we shouldn't be checking both of them the same way: the in-memory data is small and can be accessed synchronously, while the UTXO set should be queried async since it might require reading from disk.  That's the distinction I was trying to highlight.

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:23]: 👍

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:22]: 👍

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (teor2345) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:228]: Question: Are there two different kinds of timeout here?

1. If we are waiting to verify a block B, and we commit B's previous block, then we can be certain that we know the status of each UTXO in B. Each UTXO is either available now, or it will never become available.

Can we respond to all pending requests with a UTXO or error when B's parent is committed?
Should we do this?
Is it just an optimisation?

2. If we are waiting to verify a block C, but we have just started verifying from the genesis block, then it might be minutes or hours until all of C's UTXOs become available. Even if they do because available, we can't commit C until we commit all of its predecessors. In this case, we should timeout and reject C, then re-download it when we are closer to its height in the chain.

Should we timeout after some small multiple of the expected block interval?
Does the timeout here depend on C's (unverified) height?
Does the timeout depend on the re-download behaviour of the sync service?

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (teor2345) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:158]: Question: What happens when we move UTXOs from in-memory storage to sled storage?

Do we need to hold a lock over both states while we modify them?
Or is it sufficient to ensure that sled has any additional UTXOs, before we remove those UTXOs from in-memory storage?

I'd like to see this case spelled out in a bit more detail, because it's not obvious to me from the text.

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (teor2345) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:54]: This is a really important design pattern, which we might end up re-using for other data dependencies.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for this work.

I have some questions about moving UTXOs from memory to disk, and some suggestions for handling timeouts and cancellation.

How will we test this implementation?
(Should we add a testing section to all our RFCs?)

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:58]: 👍

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:65]: 🤔

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:72]: 👍

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:77]: 🙌

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:94]: 💯

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md]:
```suggestion
(mpsc) channel. The worker task receives messages and makes `Service::call`s.
```

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:133]: 👍

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (hdevalence) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:158]: I don't think that this can happen, because that movement can't happen simultaneously with checking any in-memory data. It would be a subsequent service call, after any in-memory checks had happened.

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:212]: 👍

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (dconnolly) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:227]: ☝️

**Review by dconnolly — DISMISSED** [review]
solid👍

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (hdevalence) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:228]: In case (1), I don't think that we have enough information to flush requests originating from block B with a failure, because the requests are for UTXOs by "outpoint"s (tx hash + output index), so there's not a way to trace them to a specific block as would be required.  We could add source tracking information, but I think that in this case it would be fine just to let them timeout.

In case (2), I think that it's unlikely that we'll want to start verifying C if the chain tip is very very very far behind C (e.g., all the way back to genesis).  If we did start verifying it, it would time out, but we don't remember any information about block verification failures, so there's no problem with just retrying.  The sync logic is *supposed* to restart syncing if it ever gets stuck.

I think that to start it would work to just set an unprincipled timeout (how about 120 seconds?) and then see how it works with the rest of the verification pipeline.  My guess is that we won't be timing out very often, since we'd presumably fail other checks first.

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (yaahc) [inline book/src/dev/rfcs/XXXX-asynchronous-script-verification.md]: Re above comment, what do you think about this as a clarification?

```rust
    // Calls `respond` for each (OutPoint, TransparentOutput) pair in `block`
    pub fn respond_block(&mut self, block: &Block);
```

Comment (hdevalence) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md]: I think `check` is better than `respond` because the function is checking the block against the list of pending requests, whereas `respond_block` reads as responding to the block.  It seems OK to have different names for the different functions, since they do different things: the `respond` function is called with specific response data, while `check_block` is called with a block.

Comment (yaahc) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md]: I'm okay with it not being `respond_block` but I don't think `check_block` is better, it's incredibly generic, "check_block" implies very little about what is happening to me. If anything I'd prefer to not have the helper because I think this will only end up being used one place, the entire function will boil down to just

```rust
block.transactions().iter().for_each(|transaction| {
    let hash = transaction.to_hash();
    transaction
        .outputs()
        .enumerate()
        .for_each(|(index, output)| {
            let outpoint = OutPoint { hash, index };
            self.respond(outpoint, output);
        });
});
```

we could probably add a helper to the `Block` to simplify it further

```rust
block
    .outpoints()
    .for_each(|(outpoint, output)| self.respond(outpoint, output));
```

I'm voting that we just remove this from the RFC and we can treat the helper as an implementation detail that we hash out during code review on the impl:

```suggestion
```

Comment (hdevalence) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md]: OK, sounds good!

**Review by yaahc — APPROVED** [review]
I think this is good to go once teor's comments are fully resolved, I'll leave it to them to mark it when they're satisfied.

**Comment by hdevalence** [comment]
One additional thing I'd like to nail down before we merge this is the way that we feed this information into the interface to `zcash_consensus`.  It would be unfortunate if we implemented this design and then discovered that things don't quite line up.

**Comment by hdevalence** [comment]
Blocked on #708.

File: book/src/dev/rfcs/XXXX-asynchronous-script-verification.md
Comment (teor2345) [reply book/src/dev/rfcs/XXXX-asynchronous-script-verification.md:228]: When we're at the tip, blocks should arrive every ~70 seconds.

So I think the timeout should be greater than 70 seconds, but less than some small multiple of the block interval.

That way, we don't timeout on the same block multiple times, but we do get new blocks reasonably often.

**Review by teor2345 — APPROVED** [review]
My questions have been answered.

**Comment by yaahc** [comment]
> Blocked on #708.

## Status update

I think I've resolved most of the issues in #708 other than the "windows build failure + git dependency" issue. That has been mostly finished in https://github.com/ZcashFoundation/zcashconsensus/pull/2. Right now that is set to merge into the branch we've been using to store all the work `zcashconsensus`. However that all still needs to be then merged into main, and I think there are merge conflict but I honestly have _no_ idea how they got introduced.

`zcash_script` should be ready to publish, I still need to rename the repo but I'll do that as soon as I'm done with this comment. When we do publish we need to remove the `Cargo.toml` inside the `depend/zcash` repo or the publish step will fail because it doesn't copy all the files. This is currently a bug with `cargo`, kinda, depends on who you ask, most ppl would call this a feature. **So we need to decide** if the current version is preferred (which lets us more easily track the upstream changes) or if we'd prefer to vendor the source we're pulling from `zcash` directly into our own repository. Once that question is resolved and we decide how we're going to go about publishing it (I'm partial to just publishing it personally then adding owners as necessary) it should be pretty easy to close everything up and unblock this PR.

**Comment by yaahc** [comment]
https://github.com/ZcashFoundation/zebra/pull/708 should be ready for final review and merge now

**Comment by dconnolly** [comment]
#708 has been merged, is this now unblocked or do we still have to decide on the points in https://github.com/ZcashFoundation/zebra/pull/868#issuecomment-673708142 ?

**Comment by yaahc** [comment]
@dconnolly that comment is resolved now so I think that this RFC should be good to go.

**Review by dconnolly — APPROVED** [review]

**Pull request merged**
